### PR TITLE
[RFC] Introduce avocado.core.resolver as an alternative to loader

### PR DESCRIPTION
**Pull request by clebergnu**
This new module, plugin interface and associated infrastructure
provides an alternative that is better integrated to the other Avocado
plugins.

Currently, it's only leveraged by the N(ext) runner, but plan is
to also available to the current runner.

Signed-off-by: Cleber Rosa <crosa@redhat.com>

### CODE REVIEW

**Comment by pevogam** [comment]
Hi Cleber, this are all good steps to modernize avocado's main runner, however I hope in the not so distant future the wording can be changed and this will replace the loader itself? It would be confusing for new users and anyone wishing to list tests to think whether they should load or resolve them if they are not aware of the historical background.

I find test running and loading to be two very well-established terms and if the goal of this PR is to extend loading to a more reasonable test detection I am all in for it but mostly if it actually replaces the test loading. Maybe there could be some constructs within the new loader to propose the old behavior to its users? Are there such users anyway? I am curious if people would really object the new behavior which as described in your email is way more reasonable.

**Comment by clebergnu** [comment]
> Hi Cleber, this are all good steps to modernize avocado's main runner, however I hope in the not so distant future the wording can be changed and this will replace the loader itself? It would be confusing for new users and anyone wishing to list tests to think whether they should load or resolve them if they are not aware of the historical background.
> 

I'm actively working on getting this done ASAP, so I have the same hopes you have.  With regard to the loader versus the resolution process, they should now be more distinct phases.

> I find test running and loading to be two very well-established terms and if the goal of this PR is to extend loading to a more reasonable test detection I am all in for it but mostly if it actually replaces the test loading. Maybe there could be some constructs within the new loader to propose the old behavior to its users? Are there such users anyway? I am curious if people would really object the new behavior which as described in your email is way more reasonable.

I understand "loading" as actively preparing the entity in the format that will be executed.  The resolution process is a lot more limited, and like you said, attempts to detect or check if a test seems to be available and what information we can gather on it.  Like a name resolution, resolving "www.foo.bar" may return nothing, or may return many records, but won't actually trigger a connection to that address.

So, the goal of the `resolve` command is primarily to serve as a placeholder while this feature is in development.  I agree with you that `list` is way more common and users would expect that, so we'll probably swap implementations when they are done.  The only possibility of having both at the same time is that, if for development cost reasons, we decide to hook the resolver into the nrunner only.  So until the nrunner replaces the traditional runner, we'd have both resolve (or maybe `nlist`?) just like we have `nrun` in addition to `run`.

File: avocado/plugins/resolver_avocado_instrumented.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import os
+
+from avocado.core.plugin_interfaces import Resolver
+from avocado.core.safeloader import find_avocado_tests
+from avocado.core.resolver import ReferenceResolution
+from avocado.core.resolver import ReferenceResolutionResult
+from avocado.core.nrunner import Runnable
+
+
+class AvocadoInstrumentedResolver(Resolver):
```
Comment (richtja) [inline avocado/plugins/resolver_avocado_instrumented.py]: I think the `AvocadoInstrumentedResolver`, `ExecTestResolver` and `PythonUnittestResolver` could be inside one file.  Or is there any reason to have them split into their files?

**Comment by richtja** [comment]
Hi @clebergnu I was running the `resolve` command, and when I trying to resolve some unresolved reference I was expecting some error message but there is nothing.

`avocado resolve  /not_exist
`

And when I use `--verbose` there is just list of red colored resolvers 
```
[jarichte@jarichte avocado]$ avocado resolve -V /not_exist

Resolver             Reference  Info
avocado-instrumented /not_exist
exec-test            /not_exist
glib                 /not_exist
glib                 /not_exist
python-unittest      /not_exist

```
I think it will be useful to have some error message when reference wasn't resolved at all.

File: avocado/plugins/resolver_avocado_instrumented.py (hunk 1 above)
Comment (clebergnu) [reply avocado/plugins/resolver_avocado_instrumented.py]: That's a good point.  I'll do that in the next version.

**Comment by clebergnu** [comment]
> Hi @clebergnu I was running the `resolve` command, and when I trying to resolve some unresolved reference I was expecting some error message but there is nothing.
> 
> `avocado resolve /not_exist `
> 
> And when I use `--verbose` there is just list of red colored resolvers
> 
> ```
> [jarichte@jarichte avocado]$ avocado resolve -V /not_exist
> 
> Resolver             Reference  Info
> avocado-instrumented /not_exist
> exec-test            /not_exist
> glib                 /not_exist
> glib                 /not_exist
> python-unittest      /not_exist
> ```
> 
> I think it will be useful to have some error message when reference wasn't resolved at all.

It's up to each resolver implementation to populate the "Info".  The current version I have is a bit more expressive, example:

```
$ avocado resolve --verbose /dev/nulldsadsa

Resolver             Reference       Info
avocado-instrumented /dev/nulldsadsa File path "/dev/nulldsadsa" does not end with ".py"
python-unittest      /dev/nulldsadsa File path "/dev/nulldsadsa" does not end with ".py"
glib                 /dev/nulldsadsa
robot                /dev/nulldsadsa File path "/dev/nulldsadsa" does not end with ".robot"
exec-test            /dev/nulldsadsa File "/dev/nulldsadsa" does not exist or is not a executable file
golang               /dev/nulldsadsa
```

**Comment by richtja** [comment]
> It's up to each resolver implementation to populate the "Info". 

Yes, that's,  the right way, that the resolver can specific why he can't resolve the reference. But I was thinking about some general message when reference can't be resolved by any resolver. Something like:

now:
```
[jarichte@jarichte avocado]$ avocado resolve /not_exist
[jarichte@jarichte avocado]$ 
```
with message:
```
[jarichte@jarichte avocado]$ avocado resolve /not_exist
Unable to resolve reference(s): /not_exist
[jarichte@jarichte avocado]$ 
```

**Comment by clebergnu** [comment]
Closing in favor of #3323

**Pull request closed without merging**